### PR TITLE
dont fetch coin from view 3 times when one will do.

### PR DESCRIPTION
**Pull request by Greg-Griffith**
currently we
fetch all of the tx inputs to check if spent. 
then fetch a second time to get the height to check for BIP 68
then fetch a 3rd time to get the values to calculate the fees. 

We can do this all in one step instead.

### CODE REVIEW

**Comment by ptschip** [comment]
do you have any test results such as how long does a reindex take with, without these changes?

**Comment by Greg-Griffith** [comment]
@ptschip no, i didnt think it was relevant. All this does is take the 3 for loops where we fetch Coin for an outpoint and condense it down to 1 loop so we dont keep refetching the same coin.

**Comment by Greg-Griffith** [comment]
ran reindexes as requested
dev: 2 hours 41 min.
dev + this PR: 2 hours 37 min.

File: src/validation/validation.cpp [hunk 1]
```diff
@@ -2474,32 +2474,34 @@ bool ConnectBlockCanonicalOrdering(const CBlock &block,
...
                 // be in ConnectBlock because they require the UTXO set
                 prevheights.resize(tx.vin.size());
                 {
                     for (size_t j = 0; j < tx.vin.size(); j++)
                     {
-                        prevheights[j] = CoinAccessor(view, tx.vin[j].prevout)->nHeight;
+                        CoinAccessor coin(view, tx.vin[j].prevout);
+                        // isSpend is true for empty coin object (coinEmpty)
+                        if (coin->IsSpent())
+                        {
```
Comment (ptschip) [inline src/validation/validation.cpp:2487]: It's not entirely clear to me what happens here if coin returns with a coin not found...or returns with &emptyCoin.... will that trigger coin->IsSpent?  coin->IsSpent() will be true if out.SetNull() is true but I'm not sure if that's the case here.

Comment (Greg-Griffith) [reply src/validation/validation.cpp:2487]: CoinAccessor returns either the found coin or emptyCoin. emptyCoin->IsSpent() is always true.

emptyCoin calls the default constructor for CTxOut when it creates its out datamember, and the default constructor for CTxOut calls SetNull() so IsSpent for emptyCoin should always be true.

Comment (ptschip) [reply src/validation/validation.cpp:2487]: yes that's right...ok good. will approve now.

**Review by ptschip — APPROVED** [review]

**Pull request merged**